File: app/src/main/java/org/wikipedia/feed/news/NewsItem.java
package org.wikipedia.feed.news;

import android.net.Uri;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.apache.commons.lang3.StringUtils;
import org.wikipedia.Constants;
import org.wikipedia.dataclient.WikiSite;
import org.wikipedia.dataclient.page.PageSummary;
import org.wikipedia.util.ImageUrlUtil;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public final class NewsItem {
    @SuppressWarnings("unused") @Nullable private String story;
    @SuppressWarnings("unused") @Nullable private List<PageSummary> links
            = Collections.emptyList();

    @NonNull String story() {
        return StringUtils.defaultString(story);
    }

    @NonNull public List<PageSummary> links() {
        return links != null ? links : Collections.emptyList();
    }

    @NonNull List<NewsLinkCard> linkCards(WikiSite wiki) {
        List<NewsLinkCard> linkCards = new ArrayList<>();
        for (PageSummary link : links()) {
            if (link == null) {
                continue;
            }
            linkCards.add(new NewsLinkCard(link, wiki));
        }
        return linkCards;
    }

    @Nullable public Uri thumb() {
        Uri uri = getFirstImageUri(links());
        return uri != null ? Uri.parse(ImageUrlUtil.getUrlForPreferredSize(uri.toString(), Constants.PREFERRED_CARD_THUMBNAIL_SIZE)) : null;
    }

    /**
     * Iterate through the CardPageItems associated with the news story's links and return the first
     * thumb URI found.
     */
    @Nullable private Uri getFirstImageUri(List<PageSummary> links) {
        for (PageSummary link : links) {
            if (link == null) {
                continue;
            }
            String thumbnail = link.getThumbnailUrl();
            if (thumbnail != null) {
                return Uri.parse(thumbnail);
            }
        }
        return null;
    }
}

File: app/src/main/java/org/wikipedia/feed/news/NewsItem.kt
package org.wikipedia.feed.news

import android.net.Uri
import android.os.Parcelable
import androidx.core.net.toUri
import kotlinx.parcelize.Parcelize
import kotlinx.serialization.Serializable
import org.wikipedia.Constants
import org.wikipedia.dataclient.page.PageSummary
import org.wikipedia.util.ImageUrlUtil

@Parcelize
@Serializable
class NewsItem(
    val story: String = "",
    val links: List<PageSummary> = emptyList()
) : Parcelable {
    fun thumb(): Uri? {
        return getFirstImageUri(links)?.let {
            ImageUrlUtil.getUrlForPreferredSize(
                it.toString(), Constants.PREFERRED_CARD_THUMBNAIL_SIZE
            ).toUri()
        }
    }

    fun thumbUrl(): String? {
        return getFirstImageUri(links)?.let {
            ImageUrlUtil.getUrlForPreferredSize(
                it.toString(), Constants.PREFERRED_CARD_THUMBNAIL_SIZE
            )
        }
    }

    private fun getFirstImageUri(links: List<PageSummary>): Uri? {
        return links.firstOrNull { !it.thumbnailUrl.isNullOrEmpty() }?.thumbnailUrl?.toUri()
    }
}
